src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/officialvisits/VisitSlotsMigrationService.kt

```diff
@@ -57,8 +70,121 @@ class VisitSlotsMigrationService(
+    }.onFailure {
+      failureHandler(it)
+    }.onSuccess {
+      if (it.isError) {
+        val duplicateErrorDetails = it.errorResponse!!.moreInfo
+        telemetryClient.trackEvent(
+          "nomis-migration-visitslots-duplicate",
+          mapOf(
+            "duplicateDpsId" to duplicateErrorDetails.duplicate.dpsId,
+            "duplicateNomisPrisonId" to duplicateErrorDetails.duplicate.nomisPrisonId,
+            "duplicateNomisDayOfWeek" to duplicateErrorDetails.duplicate.nomisDayOfWeek,
+            "duplicateNomisSlotSequence" to duplicateErrorDetails.duplicate.nomisSlotSequence,
+            "existingDpsId" to duplicateErrorDetails.existing.dpsId,
+            "existingNomisPrisonId" to duplicateErrorDetails.existing.nomisPrisonId,
+            "existingNomisDayOfWeek" to duplicateErrorDetails.existing.nomisDayOfWeek,
+            "existingNomisSlotSequence" to duplicateErrorDetails.existing.nomisSlotSequence,
+            "migrationId" to context.migrationId,
+          ),
+        )
+      } else {
+        telemetryClient.trackEvent(
+          "visitslots-migration-entity-migrated",
+          mapOf(
+            "nomisPrisonId" to mapping.nomisPrisonId,
+            "nomisDayOfWeek" to mapping.nomisDayOfWeek,
+            "nomisSlotSequence" to mapping.nomisSlotSequence,
+            "dpsId" to mapping.dpsId,
+            "migrationId" to context.migrationId,
+          ),
+        )
+      }
     }
   }
+
+  private suspend fun VisitTimeSlotResponse.toMigrateVisitConfigRequest(): MigrateVisitConfigRequest = MigrateVisitConfigRequest(
+    prisonCode = prisonId,
+    dayCode = dayOfWeek.asDpsApiDayOfWeek(),
+    timeSlotSeq = timeSlotSequence,
+    startTime = startTime,
+    endTime = endTime,
+    effectiveDate = effectiveDate,
+    visitSlots = visitSlots.map { visitSlot ->
+      MigrateVisitSlot(
+        agencyVisitSlotId = visitSlot.id,
+        dpsLocationId = visitSlot.internalLocation.lookUpDpsLocationId(),
+        maxGroups = visitSlot.maxGroups ?: 0,
+        maxAdults = visitSlot.maxAdults ?: 0,
+        internalLocationId = visitSlot.internalLocation.id,
+        locationKey = visitSlot.internalLocation.code,
+        maxVideoSessions = null,
+        createDateTime = LocalDateTime.now(),
+        createUsername = "TODO()",
+        modifyDateTime = null,
+        modifyUsername = null,
+      )
+    },
+    expiryDate = expiryDate,
+    createDateTime = LocalDateTime.now(),
+    createUsername = "TODO()",
+    modifyDateTime = null,
+    modifyUsername = null,
+  )
+
+  private suspend fun VisitInternalLocationResponse.lookUpDpsLocationId(): UUID = visitSlotsMappingService.getInternalLocationByNomisId(id).dpsLocationId.let { UUID.fromString(it) }
+}
+
+private fun VisitTimeSlotIdResponse.DayOfWeek.asNomisApiDayOfWeek(): VisitsConfigurationResourceApi.DayOfWeekGetVisitTimeSlot = when (this) {
```

Comment:
Starting to regret using an enum !

Reply:
Can you use this.name to grab the value - I do it for prisoner finance
postingType = GeneralLedgerEntry.PostingType.valueOf(postingType.name),